### PR TITLE
Issue 17269: formattedWrite of struct with Nullable value fails

### PR DESCRIPTION
**Pull request by CromFr**
The bug was caused by formatElement [implicitly converting](https://github.com/dlang/phobos/blob/dc07f1ae3fe1a5b1fbf8a95f39861ce87964ec58/std/typecons.d#L3136) Nullable!string to string, that resulted in [an error](https://github.com/dlang/phobos/blob/dc07f1ae3fe1a5b1fbf8a95f39861ce87964ec58/std/typecons.d#L3114) when the Nullable was null.

New behavior is to use the toString method when available, instead of implicitly converting the value.

### CODE REVIEW

File: std/format.d
Comment (atilaneves) [inline std/format.d]: Why not use a struct with `Nullable!string`?

**Comment by MetaLang** [comment]
Urgh, it's all `StringTypeOf`'s fault. That template is way too permissive.

Looks like `formatValue` uses this same method so we might as well keep it consistent. Sorry for the noise.

File: std/format.d
Comment (CromFr) [reply std/format.d]: If I use a Nullable!string, it causes an [assert failure on another unittest block](https://github.com/dlang/phobos/blob/master/std/format.d#L4035)
```
expected = `void delegate() @system`, result = `void delegate()`
```

I don't really understand why, so I wrote a custom struct to:
- Show the `alias this` and the implicit conversion problem
- Do not depend on Nullable if it is changed in the future

**Comment by CromFr** [comment]
fyi the first fix I thought of was to modify `StringTypeOf` to return false if the struct is implicitly convertible to string using an `alias this` to a function that can throw.
But since Nullable does an assert, it wouldn't work (and modifying `StringTypeOf` has too many consequences)

File: std/format.d
Comment (wilzbach) [reply std/format.d]: Huh? `Nullable!string` works without your PR: https://run.dlang.io/is/Jjzpvo

Comment (wilzbach) [reply std/format.d]: Could you provide a test case where it fails and maybe open another bug report?

Comment (CromFr) [reply std/format.d]: you need to put the Nullable!string inside a struct to see it fail:
```d
struct NullableStr {
   Nullable!string value;
}
NullableStr f;
format("%s", f).writeln;
```

**Comment by wilzbach** [comment]
Reworded the commit message, s.t. the dlang bot recognizes it and it can be auto-closed and part of the changelog (and also rebased to upstream/master).

**Comment by wilzbach** [comment]
I added the initial example of as an additional `unittest` example - this is considered a good practice.
Anyhow now we are failing due to:

```
core.exception.AssertError@std/format.d(4097): expected = `void delegate() @system`, result = `void delegate()`
```

**Comment by CromFr** [comment]
I encountered the same unittest error when initially trying to fix formattedWrite

Since it didn't make much sense to me, I changed the test to use a custom struct that has the same problem as Nullable!string and the unittest error disappeared, so I concluded that it was because of a weird import conflict or something like that, without much more investigations

**Comment by quickfur** [comment]
Rebased. Hopefully it will go through this time.

**Comment by quickfur** [comment]
Hmph. Seems that avoiding `Nullable` doesn't fix this problem. :-(  Gotta get to the bottom of that `@system delegate` issue.

**Comment by quickfur** [comment]
I think there's a compiler bug associated with this. Moving the offending unittest above the unittest that imports `Nullable` fixes the problem. Unfortunately, it's a rather elusive one, as adding `version(none)` to preceding unittests makes the problem go away. Very strange.

**Comment by CromFr** [comment]
I think dlang-bot should add a message when the PR becomes un-mergeable. I didn't receive any notification on this :/

**Comment by wilzbach** [comment]
That would be quite spammy: https://github.com/dlang-bots/dlang-bot/issues/123

Note that if the PR gets traction or approval, we as maintainers can easily rebase this.

**Comment by quickfur** [comment]
@CromFr BTW, I already rebased your pull, you may want to fetch from dlang/phobos to update your local branch if you're planning to adding more commits.

**Comment by quickfur** [comment]
Managed to reduce the bug to a minimal(-ish) case: https://issues.dlang.org/show_bug.cgi?id=18269

A truly bizarre compiler bug where unrelated declarations can affect the type of a delegate!

**Comment by MetaLang** [comment]
That's a nasty one. Thanks for the reduction @quickfur.

Anyone have a suggestion for how to proceed? I don't want this to fall by the wayside because of an obscure compiler bug.

**Comment by quickfur** [comment]
One way to deal with this, if it's deemed urgent to merge this PR sooner rather than later, is to `version` out the unittest with a comment indicating that it's being blocked by issue 18269. That way, once 18269 is fixed we can turn on the unittest again.

**Comment by wilzbach** [comment]
How about moving the unittest to a different module?

**Comment by quickfur** [comment]
On Tue, Jan 23, 2018 at 07:11:46PM +0000, Sebastian Wilzbach wrote:
>    How about moving the unittest to a different module?
[...]

I suppose we could try that too. No idea whether it will work, though.
The compiler bug seems highly sensitive and fragile to small changes.
I'd like to dig to the bottom of it sometime, but it's deep in the
complicated and messy IFTI code, and I'm not sure I'll be able to figure
it out!


--T

**Comment by quickfur** [comment]
In the meantime, perhaps we could change the order of the offending unittests so as the bypass the compiler bug?  I'm not sure when the compiler bug will be fixed; it's a pretty tricky one involving an ambiguity in the global stringtable.

**Comment by wilzbach** [comment]
Yeah that's why I thought that putting it in a separate module will help as it reduces the chances of someone accidentally running into this in the near future again.

**Comment by quickfur** [comment]
Unfortunately, the problem is cross-module. The original problem was caused by the import of `std.variant`, which in turn contains a template argument containing `void delegate()`. Basically, there's some kind of ambiguity in the compiler's stringtable, where both `void delegate()` and `void delegate() @system` have the same deco in their mangling, and whichever gets seen by the compiler first gets into the stringtable, and from that point on becomes the "official" string representation of that type. So no, putting it in a different module won't help as long as the order of unittests (as seen by the compiler) is wrong. See: https://issues.dlang.org/show_bug.cgi?id=18269

**Comment by berni44** [comment]
@RazvanN7 With #7771 fixing the problem, after fixing the merge conflict (formatspec is meanwhile scope), IMHO this can be merged now.

**Review by RazvanN7 — APPROVED** [review]

**Comment by RazvanN7** [comment]
@CromFr Please rebase this and we will get it in.

**Comment by CromFr** [comment]
Should be good now :)

**Review by RazvanN7 — APPROVED** [review]

**Comment by berni44** [comment]
@RazvanN7 circleci seems not to respond... Is there anything, that can be done?

**Comment by RazvanN7** [comment]
Superweird. I can't even access the details page.

**Comment by RazvanN7** [comment]
@berni44 I just restarted it.

**Pull request merged**
